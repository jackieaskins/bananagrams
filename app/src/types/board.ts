import { Tile } from "./tile";

export type BoardLocation = {
  x: number;
  y: number;
};

export enum Direction {
  ACROSS = "ACROSS",
  DOWN = "DOWN",
}

export enum ValidationStatus {
  NOT_VALIDATED = "NOT_VALIDATED",
  VALID = "VALID",
  INVALID = "INVALID",
}

export type WordInfo = {
  start: BoardLocation;
  validation: ValidationStatus;
};

export type AbstractBoardSquare<T> = {
  tile: T;
  wordInfo: Record<Direction, WordInfo>;
};

export type AbstractBoard<T> = Record<string, AbstractBoardSquare<T>>;

export type BoardSquare = AbstractBoardSquare<Tile>;

export type Board = AbstractBoard<Tile>;
