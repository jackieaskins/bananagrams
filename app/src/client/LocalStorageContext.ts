import { createContext, useContext } from "react";
import { SetState } from "./state/types";

type LocalStorageState<T> = [T, SetState<T>];
type LocalStorageContextState = {
  gameSettings: {
    enableTileSwap: LocalStorageState<boolean>;
  };
  savedUsername: LocalStorageState<string>;
  showTutorialPrompt: LocalStorageState<boolean>;
};

export const LocalStorageContext = createContext<LocalStorageContextState>(
  {} as LocalStorageContextState,
);

function useLocalStorageContext(): LocalStorageContextState {
  return useContext(LocalStorageContext);
}

export function useEnableTileSwap(): LocalStorageState<boolean> {
  return useLocalStorageContext().gameSettings.enableTileSwap;
}

export function useSavedUsername(): LocalStorageState<string> {
  return useLocalStorageContext().savedUsername;
}

export function useShowTutorialPrompt(): LocalStorageState<boolean> {
  return useLocalStorageContext().showTutorialPrompt;
}
