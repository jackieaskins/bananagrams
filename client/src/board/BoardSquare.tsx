import { useRecoilCallback } from 'recoil';

import { BoardPosition } from '../boards/types';
import { useCurrentBoardSquare } from '../game/stateHooks';
import { moveTile } from '../socket';
import Tile from '../tile/Tile';
import { TILE_SIZE } from '../tile/constants';
import { selectedTileState } from '../tile/selectedTileState';
import { BORDER_STYLE } from './constants';

const BoardSquare = ({ row, col }: BoardPosition): JSX.Element => {
  const square = useCurrentBoardSquare({ row, col });

  const handleClick = useRecoilCallback(
    ({ set, snapshot }) => async () => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (selectedTile) {
        moveTile({
          tileId: selectedTile.tile.id,
          fromPosition: selectedTile.boardPosition,
          toPosition: { row, col },
        });

        set(selectedTileState, null);
      }
    },
    [row, col]
  );

  const selectTile = useRecoilCallback(
    ({ set, snapshot }) => async ({ tile }) => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (!selectedTile) {
        set(selectedTileState, { tile, boardPosition: { row, col } });
      }
    },
    [col, row]
  );

  return (
    <div
      css={{
        minWidth: `${TILE_SIZE}px`,
        width: `${TILE_SIZE}px`,
        height: `${TILE_SIZE}px`,
        border: BORDER_STYLE,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
      onClick={handleClick}
    >
      {square ? (
        <Tile
          id={square.tile.id}
          letter={square.tile.letter}
          onClick={selectTile}
          validationStatus={square.validationStatus}
        />
      ) : null}
    </div>
  );
};

export default BoardSquare;
