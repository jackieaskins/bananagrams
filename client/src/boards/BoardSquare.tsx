/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import { Box } from '@material-ui/core';
import { useDrop } from 'react-dnd';

import { useStyles } from '../styles';
import { useSocket } from '../SocketContext';
import { TileItem } from '../tiles/types';
import Tile from '../tiles/Tile';
import { BoardSquare, Direction, WordInfo, ValidationStatus } from './types';

type BoardSquareProps = {
  boardSquare: BoardSquare | null;
  x: number;
  y: number;
};

type CheckValidation = (wordInfo: WordInfo) => boolean;

const getColor = (
  wordInfo: Record<Direction, WordInfo> | undefined
): string | undefined => {
  if (!wordInfo) return 'black';

  const isValid: CheckValidation = ({ validation }) =>
    validation === ValidationStatus.VALID;
  const isInvalid: CheckValidation = ({ validation }) =>
    validation === ValidationStatus.INVALID;
  const isValidated: CheckValidation = ({ validation }) =>
    validation !== ValidationStatus.NOT_VALIDATED;

  const validations = Object.values(wordInfo).filter(isValidated);

  if (validations.length === 0) return 'black';

  if (validations.every(isValid)) return 'green';
  if (validations.every(isInvalid)) return 'red';
  if (validations.some(isValid) && validations.some(isInvalid))
    return 'darkorange';

  return 'green';
};

const BoardSquare: React.FC<BoardSquareProps> = ({ boardSquare, x, y }) => {
  const classes = useStyles();
  const { socket } = useSocket();
  const { tile, wordInfo } = boardSquare ?? {};

  const [{ canDrop, isOver }, dropRef] = useDrop({
    accept: 'TILE',
    canDrop: (_, monitor) => monitor.isOver() && !tile,
    drop: ({ id, boardPosition }: TileItem) => {
      if (!!boardPosition) {
        socket.emit('moveTileOnBoard', {
          fromPosition: boardPosition,
          toPosition: { x, y },
        });
      } else {
        socket.emit('moveTileFromHandToBoard', {
          tileId: id,
          boardPosition: { x, y },
        });
      }
    },
    collect: (monitor) => ({
      isOver: monitor.isOver({ shallow: true }),
      canDrop: monitor.canDrop(),
    }),
  });

  return (
    <Box
      className={isOver && canDrop ? classes.validDrop : ''}
      height="25px"
      width="25px"
      // @ts-ignore
      ref={dropRef}
    >
      {tile ? (
        <Tile tile={tile} color={getColor(wordInfo)} boardPosition={{ x, y }} />
      ) : (
        ''
      )}
    </Box>
  );
};

export default BoardSquare;
