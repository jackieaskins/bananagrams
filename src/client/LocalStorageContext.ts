import { createContext, useContext } from "react";
import { SetState } from "./state/types";

type LocalStorageContextState = {
  gameSettings: {
    enableTileSwap: [boolean, SetState<boolean>];
  };
  savedUsername: [string, SetState<string>];
};

export const LocalStorageContext = createContext<LocalStorageContextState>(
  {} as LocalStorageContextState,
);

function useLocalStorageContext(): LocalStorageContextState {
  return useContext(LocalStorageContext);
}

export function useEnableTileSwap(): [boolean, SetState<boolean>] {
  return useLocalStorageContext().gameSettings.enableTileSwap;
}

export function useSavedUsername(): [string, SetState<string>] {
  return useLocalStorageContext().savedUsername;
}
