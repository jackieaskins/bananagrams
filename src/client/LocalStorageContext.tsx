import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { SetState } from "./state/types";

const LOCAL_STORAGE_PREFIX = "bananagrams";

type LocalStorageContextState = {
  gameSettings: {
    enableTileSwap: [boolean, SetState<boolean>];
  };
  savedUsername: [string, SetState<string>];
};

function getKey(key: string) {
  return `${LOCAL_STORAGE_PREFIX}.${key}`;
}

function useLocalStorage<T>(key: string, defaultValue: T): [T, SetState<T>] {
  const [value, setValue] = useState<T>(() => {
    try {
      return JSON.parse(
        localStorage.getItem(getKey(key)) ?? String(defaultValue),
      );
    } catch (error) {
      return defaultValue;
    }
  });

  useEffect(() => {
    localStorage.setItem(getKey(key), JSON.stringify(value));
  }, [key, value]);

  return [value, setValue];
}

const LocalStorageContext = createContext<LocalStorageContextState>(
  {} as LocalStorageContextState,
);

export function LocalStorageProvider({
  children,
}: {
  children: ReactNode;
}): JSX.Element {
  const enableTileSwap = useLocalStorage("gameSettings.enableTileSwap", false);
  const savedUsername = useLocalStorage("savedUsername", "");

  const value = useMemo(
    () => ({
      gameSettings: { enableTileSwap },
      savedUsername,
    }),
    [enableTileSwap, savedUsername],
  );

  return (
    <LocalStorageContext.Provider value={value}>
      {children}
    </LocalStorageContext.Provider>
  );
}

function useLocalStorageContext(): LocalStorageContextState {
  return useContext(LocalStorageContext);
}

export function useEnableTileSwap(): [boolean, SetState<boolean>] {
  return useLocalStorageContext().gameSettings.enableTileSwap;
}

export function useSavedUsername(): [string, SetState<string>] {
  return useLocalStorageContext().savedUsername;
}
