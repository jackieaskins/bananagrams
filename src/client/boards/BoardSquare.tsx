import { Box } from "@chakra-ui/react";
import { useDrop } from "react-dnd";
import { useEnableTileSwap } from "../LocalStorageContext";
import { useGame } from "../games/GameContext";
import Tile from "../tiles/Tile";
import { TileItem } from "../tiles/types";
import {
  BoardSquare as BoardSquareType,
  Direction,
  ValidationStatus,
  WordInfo,
} from "./types";

type BoardSquareProps = {
  boardSquare: BoardSquareType | null;
  x: number;
  y: number;
};

type CheckValidation = (wordInfo: WordInfo) => boolean;

function getColor(wordInfo: Record<Direction, WordInfo>): string | undefined {
  const isValid: CheckValidation = ({ validation }) =>
    validation === ValidationStatus.VALID;
  const isValidated: CheckValidation = ({ validation }) =>
    validation !== ValidationStatus.NOT_VALIDATED;

  const validations = Object.values(wordInfo).filter(isValidated);

  if (validations.length === 0) return "black";
  if (validations.every(isValid)) return "green.500";
  return "red.600";
}

export default function BoardSquare({
  boardSquare,
  x,
  y,
}: BoardSquareProps): JSX.Element {
  const [enableTileSwap] = useEnableTileSwap();
  const { handleMoveTileFromHandToBoard, handleMoveTileOnBoard } = useGame();
  const { tile, wordInfo } = boardSquare ?? {};

  const [{ canDrop, isOver }, dropRef] = useDrop({
    accept: "TILE",
    canDrop: (_, monitor) => monitor.isOver() && (enableTileSwap || !tile),
    drop: ({ id, boardLocation }: TileItem) => {
      if (boardLocation) {
        handleMoveTileOnBoard(boardLocation, { x, y });
      } else {
        handleMoveTileFromHandToBoard(id, { x, y });
      }
    },
    collect: (monitor) => ({
      isOver: monitor.isOver({ shallow: true }),
      canDrop: monitor.canDrop(),
    }),
  });

  return (
    <Box
      backgroundColor={isOver && canDrop ? "green.400" : undefined}
      color={isOver && canDrop ? "green.400" : undefined}
      opacity={isOver && canDrop ? 0.5 : 1}
      height="25px"
      width="25px"
      ref={dropRef}
    >
      {tile && wordInfo ? (
        <Tile tile={tile} color={getColor(wordInfo)} boardLocation={{ x, y }} />
      ) : (
        ""
      )}
    </Box>
  );
}
