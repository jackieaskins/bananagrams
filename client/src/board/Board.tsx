import { useMemo } from 'react';
import Draggable from 'react-draggable';

import BoardSquare from './BoardSquare';
import {
  BOARD_WIDTH,
  BOARD_HEIGHT,
  BORDER_STYLE,
  TILES_PER_COL,
  TILES_PER_ROW,
} from './constants';

const Board = (): JSX.Element => {
  const boardSquares = useMemo(
    () =>
      [...Array(TILES_PER_COL)].map((_, y) => (
        <div key={y} css={{ display: 'flex' }}>
          {[...Array(TILES_PER_ROW)].map((_, x) => (
            <BoardSquare key={x} />
          ))}
        </div>
      )),
    []
  );

  return (
    <Draggable bounds="parent" cancel=".no-drag">
      <div
        css={{
          cursor: 'move',
          display: 'flex',
          flexDirection: 'column',
          width: `${BOARD_WIDTH}px`,
          height: `${BOARD_HEIGHT}px`,
          border: BORDER_STYLE,
        }}
      >
        {boardSquares}
      </div>
    </Draggable>
  );
};

export default Board;
