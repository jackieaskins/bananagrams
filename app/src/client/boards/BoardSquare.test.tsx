/* eslint-disable testing-library/no-render-in-lifecycle */
import { shallow } from "enzyme";
import { useDrop } from "react-dnd";
import {
  BoardLocation,
  BoardSquare as BoardSquareType,
  Direction,
  ValidationStatus,
  WordInfo,
} from "../../types/board";
import { useEnableTileSwap } from "../LocalStorageContext";
import { boardSquareFixture, wordInfoFixture } from "../fixtures/board";
import BoardSquare from "./BoardSquare";

const mockUseEnableTileSwap = useEnableTileSwap as jest.Mock;
jest.mock("../LocalStorageContext", () => ({
  useEnableTileSwap: jest.fn().mockReturnValue([false]),
}));

const mockUseDrop = useDrop as jest.Mock;
jest.mock("react-dnd", () => ({
  useDrop: jest.fn(),
}));

const mockHandleMoveTileFromHandToBoard = jest.fn();
const mockHandleMoveTileOnBoard = jest.fn();
jest.mock("../games/GameContext", () => ({
  useGame: () => ({
    handleMoveTileFromHandToBoard: mockHandleMoveTileFromHandToBoard,
    handleMoveTileOnBoard: mockHandleMoveTileOnBoard,
  }),
}));

describe("<BoardSquare />", () => {
  const getBoardSquare = (
    across: WordInfo = wordInfoFixture(),
    down: WordInfo = wordInfoFixture(),
  ): BoardSquareType =>
    boardSquareFixture({
      wordInfo: {
        [Direction.ACROSS]: across,
        [Direction.DOWN]: down,
      },
    });

  const renderComponent = (propOverrides = {}) =>
    shallow(<BoardSquare boardSquare={null} x={0} y={0} {...propOverrides} />);

  beforeEach(() => {
    mockUseDrop.mockReturnValue([{ canDrop: true, isOver: false }, jest.fn()]);
  });

  it("renders properly when is over and can drop", () => {
    mockUseDrop.mockReturnValue([{ canDrop: true, isOver: true }, jest.fn()]);

    expect(renderComponent()).toMatchSnapshot();
  });

  it("renders properly with null boardSquare", () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  it("renders black tile when not validated in any direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.NOT_VALIDATED }),
      wordInfoFixture({ validation: ValidationStatus.NOT_VALIDATED }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  it("renders green tile when valid in every direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.VALID }),
      wordInfoFixture({ validation: ValidationStatus.VALID }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  it("renders red tile when invalid in any direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.VALID }),
      wordInfoFixture({ validation: ValidationStatus.INVALID }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  describe("useDrop", () => {
    it("is called", () => {
      renderComponent();

      expect(mockUseDrop).toHaveBeenCalledWith({
        accept: "TILE",
        canDrop: expect.any(Function),
        drop: expect.any(Function),
        collect: expect.any(Function),
      });
    });

    describe("canDrop", () => {
      const callCanDrop = (monitor) =>
        mockUseDrop.mock.calls[0][0].canDrop({}, monitor);

      it.each([
        { canDrop: true, isOver: true, swapTiles: true, tileExists: true },
        { canDrop: true, isOver: true, swapTiles: true, tileExists: false },
        { canDrop: false, isOver: true, swapTiles: false, tileExists: true },
        { canDrop: true, isOver: true, swapTiles: false, tileExists: false },
        { canDrop: false, isOver: false, swapTiles: true, tileExists: true },
        { canDrop: false, isOver: false, swapTiles: true, tileExists: false },
        { canDrop: false, isOver: false, swapTiles: false, tileExists: true },
        { canDrop: false, isOver: false, swapTiles: false, tileExists: false },
      ])(
        "returns $canDrop when monitor over is $isOver, enable tile swap is $swapTiles, and tile exists is $tileExists",
        ({ canDrop, isOver, swapTiles, tileExists }) => {
          mockUseEnableTileSwap.mockReturnValue([swapTiles]);

          renderComponent({
            boardSquare: tileExists ? boardSquareFixture() : null,
          });

          expect(callCanDrop({ isOver: () => isOver })).toBe(canDrop);
        },
      );

      it("returns true if monitor is over and no tile exists", () => {
        renderComponent();

        expect(
          callCanDrop({
            isOver: () => true,
          }),
        ).toBe(true);
      });

      it("returns false if monitor is not over but no tile exists", () => {
        renderComponent();

        expect(
          callCanDrop({
            isOver: () => false,
          }),
        ).toBe(false);
      });

      it("returns false if monitor is over and a tile exists", () => {
        renderComponent({ boardSquare: boardSquareFixture() });

        expect(
          callCanDrop({
            isOver: () => true,
          }),
        ).toBe(false);
      });

      it("returns false if monitor is not over and a tile exists", () => {
        renderComponent({ boardSquare: boardSquareFixture() });

        expect(
          callCanDrop({
            isOver: () => false,
          }),
        ).toBe(false);
      });
    });

    describe("drop", () => {
      const callDrop = (boardLocation: BoardLocation | null) =>
        mockUseDrop.mock.calls[0][0].drop({ id: "id", boardLocation });

      beforeEach(() => {
        renderComponent();
      });

      it("moves tile to new location on board if tile is already on board", () => {
        const boardLocation = { x: 1, y: 1 };
        callDrop(boardLocation);

        expect(mockHandleMoveTileOnBoard).toHaveBeenCalledWith(boardLocation, {
          x: 0,
          y: 0,
        });
      });

      it("moves tile from hand to board if tile is not on board", () => {
        callDrop(null);

        expect(mockHandleMoveTileFromHandToBoard).toHaveBeenCalledWith("id", {
          x: 0,
          y: 0,
        });
      });
    });

    describe("collect", () => {
      const mockIsOver = jest.fn().mockReturnValue("isOver");
      const mockCanDrop = jest.fn().mockReturnValue("canDrop");

      const callUseDrop = () =>
        mockUseDrop.mock.calls[0][0].collect({
          isOver: mockIsOver,
          canDrop: mockCanDrop,
        });

      beforeEach(() => {
        renderComponent();
      });

      it("calls monitor isOver", () => {
        callUseDrop();

        expect(mockIsOver).toHaveBeenCalledWith({ shallow: true });
      });

      it("calls monitor canDrop", () => {
        callUseDrop();

        expect(mockCanDrop).toHaveBeenCalledWith();
      });

      it("returns the expected valud", () => {
        expect(callUseDrop()).toEqual({
          isOver: "isOver",
          canDrop: "canDrop",
        });
      });
    });
  });
});
