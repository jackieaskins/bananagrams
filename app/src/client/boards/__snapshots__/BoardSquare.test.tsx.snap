// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardSquare /> renders black tile when not validated in any direction 1`] = `
<Box
  height="25px"
  opacity={1}
  width="25px"
>
  <Tile
    boardLocation={
      {
        "x": 0,
        "y": 0,
      }
    }
    color="black"
    tile={
      {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Box>
`;

exports[`<BoardSquare /> renders green tile when valid in every direction 1`] = `
<Box
  height="25px"
  opacity={1}
  width="25px"
>
  <Tile
    boardLocation={
      {
        "x": 0,
        "y": 0,
      }
    }
    color="green.500"
    tile={
      {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Box>
`;

exports[`<BoardSquare /> renders properly when is over and can drop 1`] = `
<Box
  backgroundColor="green.400"
  color="green.400"
  height="25px"
  opacity={0.5}
  width="25px"
/>
`;

exports[`<BoardSquare /> renders properly with null boardSquare 1`] = `
<Box
  height="25px"
  opacity={1}
  width="25px"
/>
`;

exports[`<BoardSquare /> renders red tile when invalid in any direction 1`] = `
<Box
  height="25px"
  opacity={1}
  width="25px"
>
  <Tile
    boardLocation={
      {
        "x": 0,
        "y": 0,
      }
    }
    color="red.600"
    tile={
      {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Box>
`;
